### ISSUE TITLE
Multi-file projects with individually-compilable parts, akin to LaTeX' subfiles package.

### ISSUE THREAD

**Issue by esclear** [typst/typst#2930]
### Description

This feature request originated from a question I asked on the Discord server.

I would like to implement the use case described below, but have not yet found a good way to do so.

A first idea was to create a common file `style.typ` containing the relevant set rules and show rules and include it in all files.

But this has a few problems, for example that all show-rules need to be idempotent.  
Another problem is that there can often be links or labels that refer to targets outside the subdocument, for example to a definition in another chapter.

tinger suggested the following show rule on Discord:
```typst
#show ref: it => if query(it.target, it.location()).len() == 0 [ NOT FOUND ] else { it }
```

### Use Case

I am considering writing my PhD thesis and a series of lecture notes using typst.
These will be fairly long documents, as it is often their nature.

With LaTeX, I usually do the following, using the `subfiles` package:
* Split the document into multiple, smaller files, which can be built individually.
  → These can then be distributed to others (for review, to use in preparation for the lectures, …)
* Have a "main file" that, when built, consists of all (included) parts.
  → That would be the finished thesis or the finished lecture notes at the end of the semester.

Using the `subfiles` package, all subfiles (chapters, lectures, …) use the preamble of the main file, thus providing consistent styling, etc. of both the smaller, individual documents, and the main document.  
Also, I set it up such that a corresponding cover page would automatically be generated for them.

**Comment by Dherse** [comment]
Regarding splitting your document:

- You should do it for organization
- You should not do it for performance, Typst is plenty fast and about to get much faster

As far as "phantom references" or references to "unallocated" labels, I think I *nicer* way of doing it is to have a small function like the following:

```typ
#let phantom(..keys) = {
  for key in keys.pos() {
    hide[#heading(level: 99)[no] #label(key)]
  }
}

phantom("a", "b", "c")
```

And then you can just refer to those keys, once you don't need them anymore, you can just remove them one-by-one.

**Comment by Dherse** [comment]
Yes that would indeed cause an issue with subdocuments, right now, outside of `#include`-ing all of your documents together there is no one easy way of doing that. And nor do I think it's necessary if I'm being honest, a better solution then might be to be able to render only parts of the document rather than hacking in "forward references"? Just an idea, but we could have a parameter that lets you set a section as a label `typst c my_doc.typ --section "label_name"` or something like that. I think that could be useful in general.

**Comment by Dherse** [comment]
- Internally it would need to "render" (i.e layout and introspect) all of it, it just wouldn't export all of it
- The link would probably become stale somehow

It was just an idea I threw so see whether it sticks

**Comment by laurmaedje** [comment]
> But this has a few problems, for example that all show-rules need to be idempotent.

could you elaborate what you mean with this?

> tinger suggested the following show rule on Discord:

does that suffice or are you missing something from this approach?

Overall, I see how this can in some cases be a thing you might want to do, but I'm not yet sure whether it needs any new built-in capabilities or whether it can be done fully in "user space".

**Comment by laurmaedje** [comment]
I see, that makes sense.

**Comment by memeplex** [comment]
> But this has a few problems, for example that all show-rules need to be idempotent.

I'm like a few hours new to typst so maybe this makes no sense, but if you're including sections from a master document and you have some common preamble imported and run by all of them (master+sections) in show rules, couldn't you define a state as a guard that the preamble changes the first time it runs so that it does nothing during the subsequent invocations?

Another way to decide if you are in the imported-from-master scenario or the standalone scenario may be to query for some element that only exists in the master file.

AFAICS a simpler approach like asking "is this variable defined?" for a global variable that is set from the master file and so undefined in the standalone case is not possible.

**Comment by memeplex** [comment]
Anyway, I'm in the same boat and for now I'm experimenting with a simpler approach like this (in my master file):

```typst
#let develop = true

#if develop {
  show ref: it => {
    if query(it.target, it.location()).len() == 0 [ NOT FOUND ] else { it }
  }
  include "section1.typ"
} else {
  include "section1.typ"
  include "section2.typ"
}
```

**Comment by memeplex** [comment]
I've [open a discussion](https://github.com/typst/typst/discussions/3524) about the convenience (or not) of having a more lenient compiler mode towards broken references.

**Comment by ziiirozone** [comment]
The "phantom" function @Dherse  wrote adds empty space for each key where we call the function. To solve this problem we can simply use the place function :
```typst
#let phantom(..keys) = {
  for key in keys.pos() {
    place(hide[#heading(level: 99)[no] #label(key)])
  }
}
```

Thanks for this handy function !

**Comment by joelvonrotz** [comment]
How about introducing a system variable/constant that exposes the currently targeted `.typ` file (the one you state when calling `typst compile...` or `typst watch...`). You could then check if the document you have opened is equal to the target (maybe also expose this as well to `sys`?) and then apply specific styling.

The idea is to have a more general/flexible approach to not only have a subfiles-like option.

Example:

```typst

#let subfile-template(body) = {
  // do some other styling here
  main-template(body)
}

#show: doc => if sys.target == document.filename { subfile-template(doc) } else { doc }
```

You'd define a separate template for these "subfiles" that might deal with invalid referencing. I think subfiles does something similar, where it just skips outside references.

A caviat is the `document.filename` which might not make sense in its context (as `document` might imply the target file)

I'd imagine that this would then also work in the WebApp...
